Allow all formats of dired listings

clemera opened the pull request:
Hi, thanks for this nice project! Dired also accepts custom file listings when a list is passed as argument, I added the ability to save those custom listings, too. You can test this by using something like:

    (dired
     '("*my-listing*" "~/.emacs.d" "~/.config"))

I also added storing history for dired buffers created by processes like `find-dired` or `find-grep-dired`.

clemera committed:
Allow all formats of dired listings

clemera commented:
I have found some issue with loading the history from disk, sometimes the `read` fails with the now present text properties but I haven't figured out why, yet.

clemera committed:
Avoid reusing same string object in text props

clemera commented:
Okay, fixed!

clemera committed:
Fix compiler warnings

clemera committed:
Fix indent

clemera committed:
Don't copy string object for process listings

vifon commented:
Supporting these "virtual" dired buffers wasn't initially in the scope of the project to keep things simple, though I'm not strictly against adding such support, though adding it as an opt-in feature would be welcome (controlled with a new `defcustom` variable).

As it currently is, I find your code hard to follow.  Would it be possible to break down these large `let` blocks into smaller functions?  Additionally I'm concerned about the calls such as `(current-buffer)` as the `dired-recent-path-save` function becomes much more stateful.  Did you check whether it still works with the optional argument passed?

clemera committed:
Refactor code

clemera commented:
Thanks for your feedback!

> adding it as an opt-in feature would be welcome (controlled with a new defcustom variable).

I have added the option `dired-recent-add-virtual-listings` to control this.

> Would it be possible to break down these large let blocks into smaller functions?

I have refactored the code to make it more readable.

>  Additionally I'm concerned about the calls such as (current-buffer) as the dired-recent-path-save function becomes much more stateful. Did you check whether it still works with the optional argument passed?

In master the state is captured by `default-directory` which can be wrong for virtual buffers when called from  `dired-mode-hook`. As the function is primarily intended to be called from the hook it might be okay to depend on that state. Users can now still pass any path to add one manually to the history as before.

clemera committed:
Fix docstring description

vifon commented:
Thanks for your patience. Looks good!

vifon commented:
Merged, thank you!

clemera commented:
Great, thanks!
